File: 2023/4/4-a.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func extractNumbers(input string) []int {
	fields := strings.Fields(input)
	numbers := make([]int, len(fields))
	for i, numStr := range fields {
		num := 0
		fmt.Sscanf(numStr, "%d", &num)
		numbers[i] = num
	}
	return numbers
}

func main() {
	file, err := os.Open("4-a.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var points int
	sum := 0
	for scanner.Scan() {
		points = 0

		splitByDots := strings.Split(scanner.Text(), ":")
		data := splitByDots[1]

		splitByPipe := strings.Split(data, "|")
		beforePipe := splitByPipe[0]
		afterPipe := splitByPipe[1]

		winners := extractNumbers(beforePipe)
		used := extractNumbers(afterPipe)

		fmt.Println(winners)
		fmt.Println(used)

		for _, use := range used {
			for _, winner := range winners {
				if use == winner {
					if points == 0 {
						points++
					} else {
						points *= 2
					}
				}
			}
		}
		fmt.Println("POINTS:", points)
		sum += points
	}
	fmt.Println("SUM:", sum)
}

File: 2023/4/4-a.py
sum = 0
while True:
    try:
        line = input("")
    except EOFError:
        break
    
    points = 0

    winners = line.split(":")[1].split("|")[0].split(" ")
    used = line.split(":")[1].split("|")[1].split(" ")
    
    for num in used:
        if num != '':
            if num in winners:
                if points == 0:
                    points = 1
                else:
                    points *= 2

    sum += points

print(sum)
